jetty9/src/main/java/com/linecorp/armeria/server/jetty/JettyService.java

```diff
@@ -342,63 +322,62 @@ private static void fillRequest(
-        final MetaData.Response info = transport.info;
-        if (info == null) {
-            throw new IllegalStateException("response metadata unavailable");
-        }
-
-        final ResponseHeadersBuilder headers = ResponseHeaders.builder();
-        headers.status(info.getStatus());
-        info.getFields().forEach(e -> headers.add(HttpHeaderNames.of(e.getName()), e.getValue()));
-
-        if (transport.method != HttpMethod.HEAD) {
-            headers.setLong(HttpHeaderNames.CONTENT_LENGTH, transport.contentLength);
-        }
-
-        return headers.build();
-    }
-
     private static final class ArmeriaHttpTransport implements HttpTransport {
 
-        final HttpMethod method;
-        final Queue<HttpData> out = new ArrayDeque<>();
-        long contentLength;
-        @Nullable
-        MetaData.Response info;
-        @Nullable
-        Throwable cause;
+        final HttpResponseWriter res;
 
-        ArmeriaHttpTransport(HttpMethod method) {
-            this.method = method;
+        ArmeriaHttpTransport(HttpResponseWriter res) {
+            this.res = res;
         }
 
         @Override
         public void send(@Nullable MetaData.Response info, boolean head,
                          ByteBuffer content, boolean lastContent, Callback callback) {
+            try {
+                if (info != null) {
+                    res.write(toResponseHeaders(info));
+                }
 
-            if (info != null) {
-                this.info = info;
-            }
+                if (head) {
+                    callback.succeeded();
+                    return;
+                }
 
-            final int length = content.remaining();
-            if (length == 0) {
-                callback.succeeded();
-                return;
-            }
+                final int length = content.remaining();
+                if (length == 0) {
+                    callback.succeeded();
+                    return;
+                }
 
-            if (content.hasArray()) {
-                final int from = content.arrayOffset() + content.position();
-                out.add(HttpData.wrap(Arrays.copyOfRange(content.array(), from, from + length)));
-                content.position(content.position() + length);
-            } else {
-                final byte[] data = new byte[length];
-                content.get(data);
-                out.add(HttpData.wrap(data));
+                final HttpData data;
+                if (content.hasArray()) {
+                    final int from = content.arrayOffset() + content.position();
+                    content.position(content.position() + length);
+                    data = HttpData.wrap(Arrays.copyOfRange(content.array(), from, from + length));
+                } else {
+                    final byte[] buf = new byte[length];
+                    content.get(buf);
+                    data = HttpData.wrap(buf);
+                }
+
+                if (lastContent) {
```

Comment:
Should an HTTP trailers supplied by `info.getTrailerSupplier()` be handled if `lastContent` is true?
https://github.com/eclipse/jetty.project/blob/jetty-10.0.x/jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpTransportOverHTTP2.java#L161-L173